addons/edi/models/edi_document.py

```diff
@@ -367,6 +392,52 @@ def commit_processing_state_change(self, processing_state=""):
             self.processing_state = processing_state
             self.env.cr.commit()
 
+    def commit_execute_progress(self, log_message):
+        """EDI records processing is triggered in `prepare()` and `execute()` which creates a cursor
+        savepoint to atomically commit/rollback the created/updated records. To report the progress
+        of long running executions, it is required to continually update the so-far-processed status
+        log, **while** the records are being processed. As the EDI document is locked at row-level
+        before starting process (in `lock_for_action`), writing to that record will deadlock.
+        Instead, an instance of EdiDocumentProgress will keep the progress count (which is linked to
+        an EDIDocument). The progress update logs must be committed to the DB to display the
+        progress in UI/view.
+
+        In order to commit change to 'another object' (EdiDocumentProgress in this case), create
+        another DB cursor. Transaction is committed on closing the cursor. This leaves the
+        encompassing transaction (EDI records processing) undisturbed."""
+        towrite = self.env.all.towrite.copy()
+        with self.pool.cursor() as temp_cr:
+            # NOTE: this method does a commit using another cursor, but it is called from inside a
+            # savepoint. Odoo core has a bug that it commits all pending writes without considering
+            # the cursor. To work around, clear the pending writes here; write the progress &
+            # commit; then restore those after.
+            temp_self = self.with_env(self.env(cr=temp_cr))
+            temp_self.env.all.towrite.clear()
+
+            EDIDocProgress = temp_self.env["edi.document.progress"]
+            edi_progress = EDIDocProgress.search([("doc_id", "=", temp_self.id)])
+            if not edi_progress:
+                edi_progress = EDIDocProgress.create({"doc_id": temp_self.id})
+            if edi_progress:
+                updated_log = "\n".join([edi_progress.progress_log or "", log_message])
+                try:
+                    edi_progress.write({"progress_log": updated_log})
+                except:
+                    temp_cr.rollback()
+
+        # Restore the original records to write in the main transcation
+        self.env.all.towrite = towrite
+
+    def reset_progress(self):
+        self.ensure_one()
+        EDIDocProgress = self.env["edi.document.progress"]
```

Comment:
nit: This logic is repeated a bit, maybe we could make a helper `get_or_create_edi_progress(doc_id)`

Reply:
I don't mind if this one is left, it's only repeated twice

Reply:
Yes; I was trying to avoid using `self` which has the main cursor inside `commit_execute_progress()` thus repeated the logic.